sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -3841,6 +3913,61 @@ pi_result cuda_piEnqueueMemBufferCopy(pi_queue command_queue, pi_mem src_buffer,
   }
 }
 
+pi_result cuda_piextEnqueueMemBufferCopyPeer(
+    pi_queue src_queue, pi_mem src_buffer, pi_queue dst_queue,
+    pi_mem dst_buffer, size_t src_offset, size_t dst_offset, size_t size,
+    pi_uint32 num_events_in_wait_list, const pi_event *event_wait_list,
+    pi_event *event) {
+
+  assert(src_buffer != nullptr);
+  assert(dst_buffer != nullptr);
+
+  if (!dst_queue) {
+    return PI_INVALID_QUEUE;
+  }
+
+  std::unique_ptr<_pi_event> retImplEv{nullptr};
+
+  try {
+    ScopedContext active(dst_queue->get_context());
+    if (event_wait_list) {
+      cuda_piEnqueueEventsWait(src_queue, num_events_in_wait_list,
+                               event_wait_list, nullptr);
+    }
+
+    pi_result result;
+
+    if (event) {
+      retImplEv = std::unique_ptr<_pi_event>(
+          _pi_event::make_native(PI_COMMAND_TYPE_MEM_BUFFER_COPY, dst_queue));
+      result = retImplEv->start();
+    }
+
+    auto stream = dst_queue->get();
+    auto src = src_buffer->mem_.buffer_mem_.get() + src_offset;
+    auto dst = dst_buffer->mem_.buffer_mem_.get() + dst_offset;
+
+    auto dst_context = dst_queue->get_context()->get();
+    auto src_context = src_queue->get_context()->get();
+
+    cuCtxEnablePeerAccess(src_context, 0);
```

Comment:
Is this required for doing P2P copies? My understanding was that this is to enable [direct P2P access](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PEER__ACCESS.html#group__CUDA__PEER__ACCESS_1g0889ec6728e61c05ed359551d67b3f5a) through the CUDA unified addressed pointers.

Have you tried benchmarking with and without this call? To my knowledge `cuMemcpyPeerAsync` should work either way, but maybe you're right that it needs `cuCtxEnablePeerAccess` to do enable actual P2P copying. Would be good to have it confirmed though, as otherwise you might need to [disable](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PEER__ACCESS.html#group__CUDA__PEER__ACCESS_1g5b4b6936ea868d4954ce4d841a3b4810) it somewhere as well as it appears, based on the existence of `CUDA_ERROR_TOO_MANY_PEERS`, that it is a limited resource.

Again assuming you need this call, then maybe it would be a good idea to guard it by a call to [`cuDeviceCanAccessPeer`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PEER__ACCESS.html#group__CUDA__PEER__ACCESS_1g496bdaae1f632ebfb695b99d2c40f19e).

Reply:
> Have you tried benchmarking with and without this call? To my knowledge `cuMemcpyPeerAsync` should work either way, but maybe you're right that it needs `cuCtxEnablePeerAccess` to do enable actual P2P copying. 
> 

Yes without the call to cuCtxEnablePeerAccess cuMemcpyPeerAsync performs the copy via the host.

>you might need to disable it somewhere as well as it appears, 
based on the existence of CUDA_ERROR_TOO_MANY_PEERS, that it is a limited resource.
>
Yes good idea, I am currently trying to test it with many peers on Cori.  I'm not yet sure whether it is possible with the maximum number of connections of four/eight peers in total (perlmutter/Cori cases), but probably it does no harm to call cuCtxEnablePeerAccess after cuMemcpyPeerAsync etc to cover all cases.

> Again assuming you need this call, then maybe it would be a good idea to guard it by a call to [`cuDeviceCanAccessPeer`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PEER__ACCESS.html#group__CUDA__PEER__ACCESS_1g496bdaae1f632ebfb695b99d2c40f19e).
So I believe that the call to cuDeviceCanAccessPeer is unnecessary for this use case since cuCtxEnablePeerAccess checks whether cuDeviceCanAccessPeer is false and is currently silently returning the corresponding error, before cuMemcpyPeerAsync is called anyway (making the copy via the host if cuDeviceCanAccessPeer is false).

Reply:
> Yes without the call to cuCtxEnablePeerAccess cuMemcpyPeerAsync performs the copy via the host.

Good to know! I've blindly used it in the past for this purpose, but got confused by it again because of the documentation wording. Nice to know that I wasn't being silly back then.

> Yes good idea, I am currently trying to test it with many peers on Cori. I'm not yet sure whether it is possible with the maximum number of connections of four/eight peers in total (perlmutter/Cori cases), but probably it does no harm to call cuCtxEnablePeerAccess after cuMemcpyPeerAsync etc to cover all cases.

I am not sure of the overhead incurred. Maybe it is fine to keep it enabled for future cases. What is the cost of calling `cuCtxEnablePeerAccess` on a context that already has it enabled? If it is costly, then maybe the context should be marked as having peer access enabled so the call can be skipped in subsequent P2P operations.

> So I believe that the call to cuDeviceCanAccessPeer is unnecessary for this use case since cuCtxEnablePeerAccess checks whether cuDeviceCanAccessPeer is false and is currently silently returning the corresponding error, before cuMemcpyPeerAsync is called anyway (making the copy via the host if cuDeviceCanAccessPeer is false).

As long as it returns fast and in a stable way when direct P2P support isn't possible, I agree that the `cuDeviceCanAccessPeer` would be redundant.

Reply:
> 
piextP2P is not a very informative name. Maybe something like piextDevicesSupportP2P would be more telling?
> 
 Thanks, I've made the suggested change.

Reply:
It sounds like from this [tensorflow issue](https://github.com/tensorflow/tensorflow/issues/5789) and others that when the nvidia docs say that 8 peers are supported this accounts for 8x8 peer access directions (Although there is a mention that up to 9 devices are actually supported in that tensorflow issue).  There is still a problem with connecting multiple devices on Cori so I have not been able to verify the behaviour for up to 8 devices yet. I think that if more n > 8/9 devices are used then the n*n connections will result in CUDA_ERROR_TOO_MANY_PEERS unless there are calls to cuCtxDisablePeerAccess corresponding with each call to cuCtxEnablePeerAccess.  I have experimented a little with cuCtxDisablePeerAccess and the obvious complication is that when the copy is asynchronous cuCtxDisablePeerAccess needs to wait on the memcopy call.
[LLVM OpenMP](https://reviews.llvm.org/D80649) has the same approach as used here whereby cuCtxEnablePeerAccess is called prior to every peer memory copy call without any later calls to cuCtxDisablePeerAccess.  The difference in the LLVM OpenMP approach with respect to this PR is that if Peer access is unavailable they explicitly call memcpyDtoD (which itself does the copy via the D2H H2D).  I found that this has no advantage over calling cuMemcpyPeerAsync everytime; the performance is identical.
The LLVM OpenMP implementation also outputs a debug message if cuDeviceCanAccessPeer returns false before continuing with the fallback.  I'm not sure if such a debug mode message is desired here or not?

For two devices I have confirmed that repeated calls to cuCtxEnablePeerAccess are extremely fast: for multiple cuCtxEnablePeerAccess/cuMemcpyPeerAsync calls the timing of cuCtxEnablePeerAccess was the only one output in units of nanoseconds! Another option would be to do what it sounds like tensorflow does and call cuCtxEnablePeerAccess n*n times for n devices once the contexts are created.  A problem with this is that PeerAccess can be enabled between devices when the user does not request/use the access.